### DISCUSSION TITLE
How to upgrade php version keeping backward compatibility ?

### DISCUSSION THREAD

**Discussion by beeyev** [rectorphp/rector#7520] [Q&A] [answered]
I am currently using PHP 7.0 and would like to update to PHP 7.1.
my configuration is quite simple:
```php
    $rectorConfig->phpVersion(PhpVersion::PHP_70);
    $rectorConfig->import(LevelSetList::UP_TO_PHP_71);
```

When I run rector, it makes changes which are not backward compatible with php 7.0, for example it adds public visibility prefixes to the constants. 

Is there any way to tell Rector to update the code to php 7.1 but keep it compatible with php 7.0?

**Comment by TomasVotruba** [comment, marked as answer]
Hi, 

the 2nd line changes PHP version to PHP 7.1:

```php
$rectorConfig->phpVersion(PhpVersion::PHP_70);
$rectorConfig->import(LevelSetList::UP_TO_PHP_71);
```

Change the order:

```php
$rectorConfig->import(LevelSetList::UP_TO_PHP_71);
$rectorConfig->phpVersion(PhpVersion::PHP_70);
```

**Reply by beeyev** [reply]
thank you!

**Reply by Offspring2099** [reply]
@TomasVotruba I don't understand why this setup doesn't work for me. 
I'm on php56 and moving to php70. Want the code to be compliant with php versions, until I switch the interpreter to php70. 

My config:
```
$rectorConfig->import(LevelSetList::UP_TO_PHP_70);
$rectorConfig->phpVersion(PhpVersion::PHP_56);
```

Rule `Php4ConstructorRector` gets ignored.  It has min version (provideMinPhpVersion) set to 7.0. I'm under the impression that`__constructor` style is ok in 56. Very confused.

**Reply by TomasVotruba** [reply]
I see. What would you recommend to do here?

**Reply by Offspring2099** [reply]
Hi,

I think it's correct to have `Php4ConstructorRector` included in `/config/set/php70.php`, because that was the first PHP version when the older style constructor was deemed deprecated. 

But isn't `__constructor` style constructor valid syntax in PHP5, then I would think,  `NO_PHP4_CONSTRUCTOR` should be set to some PHP5 version?

Then one can set `$rectorConfig->import(LevelSetList::UP_TO_PHP_70);` to get their code compliant with PHP70, while setting `$rectorConfig->phpVersion(PhpVersion::PHP_56);` to keep their code still compliant with PHP56 and `Php4ConstructorRector` rule would be applied and the rector'ed code should still work in PHP56 and PHP70 will not generate `e_deprecated`

EDIT:
https://wiki.php.net/rfc/remove_php4_constructors
Perhaps `NO_PHP4_CONSTRUCTOR` could be `PHP_52` (lowest php version I see in rector consts), but can go as low as php5.0.

**Reply by TomasVotruba** [reply]
Hi, the upgrade sets follow "upgrade-early" approach. Otherwise, project would be in a very bad shape even in PHP 9.
So the rule should be in the set it was the first introduced :+1:

**Reply by Offspring2099** [reply]
Sorry if I'm completely misunderstanding this, I'm not disputing that the set is wrong. The 70 set is correct to include `Php4ConstructorRector` rule. 

It seems that `Php4ConstructorRector` rule is restricted, not to upgrade, unless the current php version is 70.  Which does seem to contradict the "upgrade-early" approach. 
In the case of this config:
```
$rectorConfig->import(LevelSetList::UP_TO_PHP_70);  <--- want my code to be php70 ready
$rectorConfig->phpVersion(PhpVersion::PHP_56);     <---- but I'm still on php56
```
Wouldn't one want the rule applied in php5 (that would be earlier than 7)

**Reply by TomasVotruba** [reply]
> It seems that Php4ConstructorRector rule is restricted, not to upgrade, unless the current php version is 70. Which does seem to contradict the "upgrade-early" approach.

That's what should be fixed then :+1:

**Reply by themasch** [reply]
Sorry to dig up old discussions here, but I just run into a similar problem with the 8.4 upgrade and explicit nullables.
The `ExplicitNullableParamTypeRector` is gates behing PHP_84, but the resulting code is actually valid since PHP 7.1. 
There where attempts to address this, but it was decided that it should remain that way: 
See for example this: https://github.com/rectorphp/rector-src/pull/5848
And this discussion: https://github.com/rectorphp/rector-src/pull/5724#pullrequestreview-2029763095

For me the `minPhpVersion` currently attempts to do two things at once: For some rules, it definies the actual PHP version where the result becomes valid PHP code, e.g. `AddTypeToConstRector` is gates behint PHP_83 because before that its not valid. 
For `ExplicitNullableParamTypeRector` on the other hand the `minPhpVersion` describes the PHP version where the *old* code is no longer "good" (e.g. deprecated). 

So currently theres not way to "upgrade early", e.g. get 8.4 compliant code that also works fine with 8.3, except setting `$rectorConfig->phpVersion(PhpVersion::PHP_84)` and hoping it doesnt attempt to generate code that actually _requires_ 8.4 (dropping parantheses on constructors, property hooks, etc). 

Has the projects view on this "upgrade-early" approach changed?

**Reply by samsonasik** [reply]
@themasch see answered question regarding `ExplicitNullableParamTypeRector` https://github.com/rectorphp/rector/discussions/8823#discussioncomment-10599155

**Reply by themasch** [reply]
@samsonasik thanks for the quick response, I was not aware that implicit nullables behaved differently regarding that optional-paramter-before-required-paramter rule. 
That indeed makes it harder to generally apply that rector to code pre 8.4. 
Makes sense now, thanks!

**Comment by camlafit** [comment]
Hi

Sorry to reopen this discussion.


With last rector version, this answers doesn't look anymore accurate. What is the (new) correct way ?
As is legacy code I can't set version php in any composer.json and php use in dev is different from the prod system.

I've tried :
```php
    ->withSets([LevelSetList::UP_TO_PHP_82])
    ->withPhpVersion(PhpVersion::PHP_74)
```
or 
```php
    ->withPhpVersion(PhpVersion::PHP_74)
    ->withSets([LevelSetList::UP_TO_PHP_82])
```
without success.

I've tried also with ```->withPhpSets(php74:true)```

If i add ``` int | bool $my_variable = 1;``` any error is raise.

Looks working with 
```php
    ->withDowngradeSets(false, false, false, false, false,true)
    ->withPhpVersion(PhpVersion::PHP_74)
```

But in this case it's a full downgrade until php7.4 , If I've understood correctly downgrade approach don't target interval compatibility. I'm wrong ?
